### PR TITLE
BumbleBee_WendyWang

### PR DESCRIPTION
**Pull request by wendyww9**

### CODE REVIEW

File: app/db.py [hunk 1]
```diff
@@ -3,4 +3,4 @@
 from .models.base import Base
 
 db = SQLAlchemy(model_class=Base)
-migrate = Migrate()
```
Comment (kelsey-steven-ada) [inline app/db.py]: It looks like only a blank line at the end of the file was removed here. If we aren't making meaningful changes to a file, we typically don't want to check that into version control. When we are committing changes, we want to review the files first to ensure we aren't committing a change like this that we didn't necessarily intend.

File: app/models/goal.py [hunk 2]
```diff
@@ -1,5 +1,20 @@
...
 class Goal(db.Model):
     id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
+    title: Mapped[str]
+    tasks: Mapped[list["Task"]] = relationship(back_populates="goal")
+
+    def to_dict(self):
+        return {
+            "id": self.id,
+            "title": self.title,
+            #"tasks": [task.to_dict() for task in self.tasks]
```
Comment (kelsey-steven-ada) [inline app/models/goal.py]: Reviewing our code for comments or testing code left behind should be one of our final steps in completing a coding task. Commented code should be removed prior to opening PRs so it does not get added to the production code base.

File: app/models/task.py [hunk 3]
```diff
@@ -1,5 +1,38 @@
...
 from ..db import db
-
+from datetime import datetime
+from typing import Optional
+from sqlalchemy import ForeignKey
 class Task(db.Model):
     id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
+    title: Mapped[str]
+    description: Mapped[str]
+    completed_at: Mapped[datetime | None] = mapped_column(default=None, nullable=True)
```
Comment (kelsey-steven-ada) [inline app/models/task.py]: We should use one style for how we declare an attribute at optional across all of our files within a project and stay consistent. 
- If you were to stay with the `Mapped[datetime | None]` style syntax, we should remove `Optional` and use just that syntax. 
- Conversely, we could remove the `Mapped[datetime | None]` style syntax, and stick with just `Optional` but we should not be mixing both styles in the same project.

Since you imported and used `Optional` elsewhere, another way we could declare this column is:
```py
completed_at: Mapped[Optional[datetime]]
```

I'd need to test this, but with the current syntax `completed_at: Mapped[datetime | None] = mapped_column(default=None, nullable=True)`, I believe `Mapped[datetime | None]` is enough to mark the attribute as optional, you shouldn't need to also declare `nullable=True` or `default=None` in the `mapped_column` invocation.

File: app/models/task.py [hunk 4]
```diff
@@ -1,5 +1,38 @@
...
+    goal_id: Mapped[Optional[int]] = mapped_column(ForeignKey("goal.id"))
+    goal: Mapped[Optional["Goal"]] = relationship(back_populates="tasks")
+
+    @property
+    def is_complete(self):
+        return self.completed_at is not None
```
Comment (kelsey-steven-ada) [inline app/models/task.py:17]: Nice way to keep the logic for `is_complete` bundled with the model without duplicating any information!

File: app/models/task.py [hunk 5]
```diff
@@ -1,5 +1,38 @@
...
+        return task_as_dict
+    
+    @classmethod
+    def from_dict(cls, task_data):
+        completed_at = task_data.get("completed_at", None)
+        return cls(
+            title=task_data["title"],
+            description=task_data["description"],
+            completed_at=completed_at,
+            goal_id=task_data.get("goal_id", None)
```
Comment (kelsey-steven-ada) [inline app/models/task.py]: `.get` is used to return a value or `None` for both potentially missing fields, but they are handled slightly differently. Is there a reason to pull `completed_at` out into it's own variable, or could we follow the same pattern as you used for `goal_id`?

File: migrations/versions/719d6486b24f_.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
+"""empty message
```
Comment (kelsey-steven-ada) [inline migrations/versions/719d6486b24f_.py:1]: Just like git commits, we should enter a message with every migration that describes the changes so if someone needs to look back at the history of database changes they have context without needing to read through the whole file's code.

File: test.sh [hunk 7]
```diff
@@ -1,31 +1 @@
...
-        echo "Timed Out - Database not Found"
-        exit 1
-      fi
-
-    done
-fi
-
-pytest
\ No newline at end of file
+echo "hello world"
```
Comment (kelsey-steven-ada) [inline test.sh]: Was this change intentional? 
If so, what was the purpose of removing this script file?

If you want to create your own file for testing things out, that is always welcome. This script is not part of the project instructions, and we should not change parts of the project scaffold that are not necessary to be updated or deleted to complete our work.

File: tests/test_wave_01.py [hunk 8]
```diff
@@ -107,17 +107,17 @@ def test_update_task(client, one_task):
...
-
+    
     assert task.title == "Updated Task Title"
     assert task.description == "Updated Test Description"
     assert task.completed_at == None
 
 
 
-@pytest.mark.skip(reason="No way to test this feature yet")
+###@pytest.mark.skip(reason="No way to test this feature yet")
```
Comment (kelsey-steven-ada) [inline tests/test_wave_01.py]: Comment formatting
```suggestion
#@pytest.mark.skip(reason="No way to test this feature yet")
```

File: app/routes/goal_routes.py [hunk 9]
```diff
@@ -1 +1,90 @@
-from flask import Blueprint
\ No newline at end of file
+from flask import Blueprint, abort, make_response, Response, request
+from ..db import db
+from app.models.goal import Goal
+from app.models.task import Task
+from .route_utilities import validate_model, create_model
+import requests
+import json
```
Comment (kelsey-steven-ada) [inline app/routes/goal_routes.py]: I'm not seeing where this import is used. If an import is not required, we should remove it so that our files are not loading up extra code into our execution environment that will never be used. This makes our programs larger and run slower without any benefit.

This feedback on this import applies to the `task_routes.py` file as well.

File: app/routes/goal_routes.py [hunk 10]
```diff
@@ -1 +1,90 @@
...
+import os
+
+bp = Blueprint("goals_bp",__name__,url_prefix="/goals")
+
+@bp.post("")
+def create_goal():
+    request_body = request.get_json()
+    data, status_code = create_model(Goal, request_body)
+    return {"goal": data}, status_code
```
Comment (kelsey-steven-ada) [inline app/routes/goal_routes.py]: Really nice create function, great use of helpers to make this clear and concise!

File: app/routes/goal_routes.py [hunk 11]
```diff
@@ -1 +1,90 @@
...
+@bp.get("")
+def get_all_goals():
+    query = db.select(Goal)
+    goals = db.session.scalars(query)
+
+    goals_response = []
+    for goal in goals:
+        goals_response.append(goal.to_dict())
+    
+    return goals_response, 200
```
Comment (kelsey-steven-ada) [inline app/routes/goal_routes.py]: In Python, it would be more common and preferred to use a comprehension here:
```py
goals_response = [goal.to_dict() for goal in goals]
return goals_response 
```

A note on the code above: 200 is the default status code for routes in Flask and will be passed as the status code if we do not provide one explicitly.

File: app/routes/goal_routes.py [hunk 12]
```diff
@@ -1 +1,90 @@
...
+@bp.get("/<goal_id>/tasks")
+def get_all_goal_tasks(goal_id):
+    goal = validate_model(Goal, goal_id)
+    tasks = [task.to_dict() for task in goal.tasks]
+
+    return {
+        "id": goal.id,
+        "title": goal.title,
+        "tasks": tasks
+    }, 200
```
Comment (kelsey-steven-ada) [inline app/routes/goal_routes.py]: Looking across the routes that should return a 200 status code, I see that sometimes the 200 status code is explicitly added in the return, and sometimes it is not. We should either always explicitly add the status code, or always remove it if we are returning the default 200, so that we are handling the situation consistently across the entire project.

File: app/routes/goal_routes.py [hunk 13]
```diff
@@ -1 +1,90 @@
...
+    return goals_response, 200
+
+
+@bp.get("/<goal_id>")
+def get_one_goal(goal_id):
+    goal = validate_model(Goal, goal_id)
+    
+    return {"goal": goal.to_dict()}
+
+@bp.get("/<goal_id>/tasks")
```
Comment (kelsey-steven-ada) [inline app/routes/goal_routes.py]: Spacing between functions should be consistent across a file. In Python, the PEP8 recommendation is 2 blank lines between functions that are not contained within a class.

File: app/routes/goal_routes.py [hunk 14]
```diff
@@ -1 +1,90 @@
...
+        "title": goal.title,
+        "tasks": tasks
+    }, 200
+
+
+@bp.put("/<goal_id>")
+def update_one_goal(goal_id):
+    goal = validate_model(Goal, goal_id)
+    request_body = request.get_json()
+    goal.title = request_body['title']
```
Comment (kelsey-steven-ada) [inline app/routes/goal_routes.py]: To keep stronger separation of concerns, just like we have a `to_dict` and `from_dict` function, we could have an `update` function on the model class. That way, the routes themselves don't need to have knowledge of a model's specific attributes.

File: app/routes/goal_routes.py [hunk 15]
```diff
@@ -1 +1,90 @@
...
+bp = Blueprint("goals_bp",__name__,url_prefix="/goals")
+
+@bp.post("")
+def create_goal():
+    request_body = request.get_json()
+    data, status_code = create_model(Goal, request_body)
+    return {"goal": data}, status_code
+
+@bp.post("/<goal_id>/tasks")
+def create_task_with_goal_id(goal_id):
```
Comment (kelsey-steven-ada) [inline app/routes/goal_routes.py]: We are not creating any Task instances in this route, so how could we rename the function to better reflect the actions it is taking?

File: app/routes/goal_routes.py [hunk 16]
```diff
@@ -1 +1,90 @@
...
+    goal = validate_model(Goal, goal_id)
+    request_body = request.get_json()
+    
+    task_ids = request_body.get("task_ids")
+
+    for task in goal.tasks:
+        task.goal_id = None
```
Comment (kelsey-steven-ada) [inline app/routes/goal_routes.py]: Another approach to breaking the association between goals and tasks is to set `goal.tasks` to an empty list, that way we don't need to iterate through all the associated tasks:
```py
goal.tasks = []
```

File: app/routes/goal_routes.py [hunk 17]
```diff
@@ -1 +1,90 @@
...
+    
+    tasks = []
+    for task_id in task_ids:
+        task = validate_model(Task, task_id)
+        task.goal_id = goal.id
+        tasks.append(task)
```
Comment (kelsey-steven-ada) [inline app/routes/goal_routes.py:28]: It is functionally about the same, but another approach could be to create our list of validated tasks, and then assign it to `goal.tasks`:
```py
valid_tasks = [validate_model(Task, task_id) for task_id in task_ids]
goal.tasks = valid_tasks
```

File: app/routes/goal_routes.py [hunk 18]
```diff
@@ -1 +1,90 @@
...
+
+@bp.get("/<goal_id>/tasks")
+def get_all_goal_tasks(goal_id):
+    goal = validate_model(Goal, goal_id)
+    tasks = [task.to_dict() for task in goal.tasks]
+
+    return {
+        "id": goal.id,
+        "title": goal.title,
+        "tasks": tasks
```
Comment (kelsey-steven-ada) [inline app/routes/goal_routes.py]: We are duplicating some effort here, `Goal` already has a `to_dict` function that creates a dict with the `id` and `title`. To D.R.Y up the code:
- we could add a parameter to the `to_dict` function to toggle whether we want to add the tasks info 
- we could create a new instance function on the model that calls `to_dict` and then adds the tasks

Either way would let us reuse the code that already exists, and keep information about the model attributes in the model class so that our routes don't need to have special knowledge.

File: app/routes/route_utilities.py [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+        abort(make_response(response, 404))
+
+    return model
+
+def create_model(cls, model_data):
+    try:
+        new_model = cls.from_dict(model_data)
+    except KeyError as e:
+        response = {"details": "Invalid data"}
+        abort(make_response(response, 400))
```
Comment (kelsey-steven-ada) [inline app/routes/route_utilities.py]: If we aren't reading values from the error object, we would typically leave off the `as e` portion of the `except`. 

I also strongly encourage using a longer, more descriptive, name than `e` to represent an error. Given that it's a key error that arises from trying to create an instance, what would be a more descriptive name for the error?

If we wanted to incorporate specific information about what key caused the KeyError in the message, we could do that with something like:
```py
except KeyError as error:
    response = {"details": f"Invalid data: missing {error.args[0]}"}
    abort(make_response(response, 400))
```

File: app/routes/route_utilities.py [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+
+#     if filters:
+#         for attribute, value in filters.items():
+#             if hasattr(cls, attribute):
+#                 query = query.where(getattr(cls, attribute).ilike(f"%{value}%"))
+    
+#     models = db.session.scalars(query.order_by(cls.id))
+#     models_response = [model.to_dict() for model in models]
+
+#     return models_response
```
Comment (kelsey-steven-ada) [inline app/routes/route_utilities.py]: We should removed unused or commented code.

File: app/routes/route_utilities.py [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+    query = db.select(cls).where(cls.id == model_id)
+    model = db.session.scalar(query)
+
+    if not model:
+        response = {"message": f"{cls.__name__} with id {model_id} does not exist"}
+        abort(make_response(response, 404))
+
+    return model
+
+def create_model(cls, model_data):
```
Comment (kelsey-steven-ada) [inline app/routes/route_utilities.py:21]: Since this both creates a model and creates an http response, I'd consider renaming this to `create_model_and_response`. Or to better follow single responsibility principles, we could split this into 2 functions:
- one function that only manages creating and returns the new model
- one function that uses the function above to get the model, then creates and returns the response based on that model.

File: app/routes/task_routes.py [hunk 22]
```diff
@@ -1 +1,102 @@
...
+
+    return Response(status=204, mimetype ="application/json")
+    
+@bp.patch("/<task_id>/mark_complete")
+def update_one_task_complete(task_id):
+    task = validate_model(Task, task_id)
+    task.completed_at = datetime.now()
+    
+    db.session.commit()
+    db.session.commit()
```
Comment (kelsey-steven-ada) [inline app/routes/task_routes.py]: One of these duplicated lines should be removed so we only call `db.session.commit()` once.

File: app/routes/task_routes.py [hunk 23]
```diff
@@ -1 +1,102 @@
...
+    call_slack_api(task.title)
+
+    return Response(status=204, mimetype ="application/json")
+
+def call_slack_api(title):
```
Comment (kelsey-steven-ada) [inline app/routes/task_routes.py:76]: Nice separation of concerns and sticking to the single responsibility principle by breaking out sending the Slack message into its own function!

**Review by kelsey-steven-ada — COMMENTED** [review]
Nice work! Please review the feedback and questions in the comments when you have time. Let me know here or on Slack if you have questions or there's anything I can clarify =]

File: app/models/goal.py (hunk 2 above)
Comment (wendyww9) [reply app/models/goal.py]: Got it.

File: migrations/versions/719d6486b24f_.py (hunk 6 above)
Comment (wendyww9) [reply migrations/versions/719d6486b24f_.py:1]: Got it.

File: test.sh (hunk 7 above)
Comment (wendyww9) [reply test.sh]: Sorry, it was for something else. It should not be here.

File: tests/test_wave_01.py (hunk 8 above)
Comment (wendyww9) [reply tests/test_wave_01.py]: Got it.

File: app/routes/goal_routes.py (hunk 11 above)
Comment (wendyww9) [reply app/routes/goal_routes.py]: Will practice more about comprehension.

File: app/routes/goal_routes.py (hunk 13 above)
Comment (wendyww9) [reply app/routes/goal_routes.py]: Will check PEP8 for styling.

File: app/models/task.py (hunk 5 above)
Comment (wendyww9) [reply app/models/task.py]: > `.get` is used to return a value or `None` for both potentially missing fields, but they are handled slightly differently. Is there a reason to pull `completed_at` out into it's own variable, or could we follow the same pattern as you used for `goal_id`?

completed_at can follow the same pattern as goal_id.

File: app/routes/goal_routes.py (hunk 15 above)
Comment (wendyww9) [reply app/routes/goal_routes.py]: assign_tasks_to_goal_id

File: app/routes/goal_routes.py (hunk 17 above)
Comment (wendyww9) [reply app/routes/goal_routes.py:28]: Thanks. But then we still have to write a for loop to assign task.goal_id?

File: app/routes/route_utilities.py (hunk 19 above)
Comment (wendyww9) [reply app/routes/route_utilities.py]: can use error.

File: app/routes/goal_routes.py (hunk 17 above)
Comment (kelsey-steven-ada) [reply app/routes/goal_routes.py:28]: This is one of the the cool things about using an ORM like SQLAlchemy and creating the convenience attributes using `relationship`! 

By assigning the `goal` instance's `tasks` attribute to a list of `Task` objects, SQLAlchemy knows that we are associating those models using the one to many relationship we established with foreign keys. As long as we call `db.session.commit()` after updating `goal.tasks`, the `goal_id` fields will be properly updated on the Tasks.
